left_sidebar: Fix DM row highlight linger when view is changed.

Rinwaoluwa opened the pull request:
**Description**
This work aims to fix Dm row  linger, Previously when a user navigates away from Direct messages view e.g clicking on a Stream the Direct messages view still remains open.  

Now when a user navigates to other parts of the UI that is not a Direct Message the Direct messages view is collapsed.


**Before**



https://github.com/zulip/zulip/assets/117735533/8b2b588a-fcc3-4592-9e65-69531f9ae0f9




**After**




https://github.com/zulip/zulip/assets/117735533/b7dd467a-a2ff-4ea4-99c9-5c768c93ebc7



Fixes https://github.com/zulip/zulip/issues/27698
<details>
<summary>Self-review checklist</summary>




- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

shashank-23002 commented on web/src/left_sidebar_navigation_area.js:
```diff
@@ -154,6 +155,7 @@ function do_new_messages_animation($li) {
 }
 
 export function highlight_inbox_view() {
+    pm_list.deselect_dm_list_item();
```
Instead of using the ```deselect_dm_list_item()``` function , I think you should have a look at ```close()``` function in ```pm_list.js``` as it is what the issue needs to resolve.
```suggestion
    pm_list.close();
```

Rinwaoluwa replied:
@shashank-23002 Thank you so much for your review, As i am looking in [CZO](https://chat.zulip.org/#narrow/stream/6-frontend/topic/lagging.2C.20lingering.20DM-row.20highlight) most of the conversations revolves around the highlight linger, I might have misunderstood something somewhere. Thank you so much for the Review once again I'll work on my patch and revert back ASAP.

Rinwaoluwa replied:
@shashank-23002 Thank you so much for the Tip, I have looked into the close function and I do see that it is what i need, my question is, do i call it only when narrowed to a stream or when any other row in the leftside bar is clicked

shashank-23002 reviewed (commented):
Hey @Rinwaoluwa , thanks for working on the issue. The changes looks good to me if the issue required fixing just the highlight linger part. However this PR needs some little changes - 

1) The commit message does not follow the commit guidelines as it requires the ```Fixes:__``` part. 
2) If you could please update the screenshots section to shift above the self-review checklist part. As you can look in the preview section the screenshots are not visible until you open the checklist section.
3) This issue requires us to fix the DM row linger issue and not specifically the highlight part (it is a small part of issue) which occurs when the user clicks on some other part of DOM. 

Would be great if you could clearly look upon the issue description and change accordingly.

N-Shar-ma reviewed (changes requested):
While this does work, it would be nice if you could make a clean up commit too before this to use the `deselect_top_left_corner_items` function instead of removing highlights individually in functions like `highlight_recent_view`.

Then for this main commit you could either include a call to the new `deselect_dm_list_item` in `deselect_top_left_corner_items` (renaming it accordingly) or make a new functioning compositing these 2, and use that instead, to reduce duplication.

Rinwaoluwa commented:
@N-Shar-ma Thank you so much for your review, I'll get to work right now.

Rinwaoluwa commented:
Hi @N-Shar-ma , I have worked on the PR 

1. `deselect_top_left_corner_items` now removes highlights.
2. while refactoring I noticed  `deselect_top_left_corner_items` was exported but I did a `git grep` and found that the function wasn't used anywhere outside of the file so i took out the export in another commit.
3. I added `deselect_dm_list_item` into `deselect_top_left_corner_items` and renamed `deselect_top_left_corner_items` to `deselect_left_corner_items`.

N-Shar-ma commented:
@Rinwaoluwa Those commits look fine to me! Just rename the composite function to something clearer like `deselect_non_strteam_left_sidebar_items`, and this should be good to go from my end

Rinwaoluwa commented:
@N-Shar-ma I renamed the function, Thank you for your review.

alya commented:
@timabbott This was reviewed by @N-Shar-ma . Let me know if you'd like me to try testing it.

Rinwaoluwa commented:
Hi @timabbott, this PR still needs your response to move forward. Thank you.

timabbott commented on web/src/left_sidebar_navigation_area.js:
```diff
@@ -85,7 +87,7 @@ function deselect_top_left_corner_items() {
 }
```
How do stream items get deselected? I would kinda prefer to keep the DM implementation parallel to that, rather than paralleling the top-left items.

Rinwaoluwa replied:
would you prefer for me to add the `deselect_stream_items` there ?

timabbott commented:
The first two commits seem like good refactors; I'm not sure the last is the cleanest approach -- can you talk about what alternatives you considered and why you picked this approach?

Rinwaoluwa commented:
> The first two commits seem like good refactors; I'm not sure the last is the cleanest approach -- can you talk about what alternatives you considered and why you picked this approach?

- I thought about calling `deselect_dm_list_item` in `stream_list.js` and`left_sidebar_navigation_area.js` but that would be too much duplicate code.
- I also thought of moving the `is_active` logic in the templates file (this is adds the highlight) into `pm_list.js ` to have all the logic of highlighting a `dm_list_item` but this creates a bug where no `dm_list_item` gets highlighted.
- I chose this approach because it works and i understand how and why it does, coming from a `dm_list_item` if any other part of the left sidebar UI i.e. Not a `dm_list_item` is clicked it will be highlighted in parallel with `dm_list_item`.

NB: being that there is no `deselect_dm_list_item` to remove the highlight i wrote one to remove the `active sub filter ` class from the highlighted  `dm_list_item`.
1. Because `deselect_dm_list_item` strips away the `active sub filter ` class from the highlighted `dm_list_item` I wrote a `highlight_dm_list_item` to add the class back.
2. Calling `deselect_dm_list_item` inside the `left_sidebar_navigation_area.js` file makes it so that when any of the top left corner items is clicked the `dm_list_item` is deselected.

timabbott commented:
How do we ensure that topics/streams in the left sidebar get deselected when navigating away from those narrows? I would have expected the DM code to want to do something very parallel to that.

Rinwaoluwa commented:
> How do we ensure that topics/streams in the left sidebar get deselected when navigating away from those narrows? I would have expected the DM code to want to do something very parallel to that.

Thank you for the review Tim, I was able to reconsider my approach and make `deselect_dm_list_item` work in parallel with stream items deselection. I am ready for another review.

timabbott commented:
I've confirmed this does not fix https://github.com/zulip/zulip/issues/27698 entirely -- it only fixes the highlighting bug. There's still the bug that the row remains displayed even though "DIRECT MESSAGES" is collapsed. I think `update_private_messages` is supposed to do that? 

I made a few naming/ordering tweaks, adjusted the message to not incorrectly close the issue, and pushed this back here as a potentially useful checkpoint.

But I think actually we should debug why `update_private_messages` is not emptying the section.

Rinwaoluwa commented:
> I've confirmed this does not fix #27698 entirely -- it only fixes the highlighting bug. There's still the bug that the row remains displayed even though "DIRECT MESSAGES" is collapsed. I think `update_private_messages` is supposed to do that?
> 
> I made a few naming/ordering tweaks, adjusted the message to not incorrectly close the issue, and pushed this back here as a potentially useful checkpoint.
> 
> But I think actually we should debug why `update_private_messages` is not emptying the section.

I do know why the row remains displayed even though "DIRECT MESSAGES" is collapsed, but I didn't think that was a bug. I think it's a feature because in `update_private_messages` there's an if statment for `private messages collapsed` inside the block it states that if there is an active conversation it should be keyed even when "DIRECT MESSAGES" is collapsed.

timabbott commented:
But isn't the whole bug that we're leaving the direct message conversation view and navigating to a stream view, and the conversation remains listed?

Rinwaoluwa commented:
I agree with you, I'll work on it right now.

Rinwaoluwa commented:
@timabbott I'm ready for another review.

Rinwaoluwa commented:
@timabbott  you haven't reviewed this!

timabbott commented:
This PR doesn't look ready for a review from me; I see various typos both in the code comments and commit messages, unintended changes to whitespace, etc. Please clean up your PR, including a longer write-up about how you've addressed the feedback and how you've verified that this version is correct and doesn't introduce new bugs. Maintainers do not have time to review changes that have problems like this on a cursory look -- it generally indicates sloppyness in the work itself.

Rinwaoluwa commented:
Thank you Tim for the review. I've worked on the PR as requested and addressed the feedback.

timabbott commented:
The last commit message still mentions a function that doesn't exist. Merged the two other commits as fbe7285af8163fb2e2ea0afb36a6645512bb779e.

timabbott commented on web/src/pm_list.js:
```diff
@@ -33,7 +36,6 @@ export function close() {
     private_messages_collapsed = true;
     $("#toggle_private_messages_section_icon").removeClass("fa-caret-down");
     $("#toggle_private_messages_section_icon").addClass("fa-caret-right");
-
```
Don't remove blank lines in code not touched by your PR. It's distracting for reviewers.

timabbott commented on web/src/pm_list.js:
```diff
@@ -141,6 +152,13 @@ function scroll_all_private_into_view() {
     const active_filter = filter;
+    is_private_messages_in_view = active_filter.has_operator("dm");
+    unhighlight_dm_list_item();
+    /* Close private message section on user exit*/
```
The whitespace in this comment is wrong.

Rinwaoluwa commented:
@timabbott I'm ready for another review, Thank you.

Rinwaoluwa commented:
@timabbott, This still needs your review.

Rinwaoluwa commented:
I tried reporting this failed test in [CZO](https://chat.zulip.org/#narrow/stream/43-automated-testing/topic/puppeteer.20tests.20are.20failing.20at.20download/near/1743238) but as @timabbott  said **the third-party server hosting those files for the puppeteer package is broken.**  I figured my PR isn't breaking anything. I fixed the conflicting changes.

Rinwaoluwa commented:
@timabbott I've resolve the conflicts and i'm ready for another review.

Rinwaoluwa commented:
@timabbott This still needs your review please.

pratik-pc commented on web/src/pm_list.ts:
```diff
@@ -144,6 +156,18 @@ function scroll_all_private_into_view(): void {
+        /*
+            Since handle_narrow_activated is called within
+            narrow.activate, and narrow.activate is responsible
+            for building narrows, this is the most efficient place
+            to handle user navigation away from a DM view.
+        */
+        close();
```
Is this necessary since it will change the current design by collapsing DM every time user navigates away. Maybe we could just use `update_private_messages` directly

alya replied:
That behavior is definitely undesirable.

Rinwaoluwa replied:
Hi @alya, @pratik-pc 

Thank you both for taking the time to review my pull request. I appreciate your feedback.

However, I'd like to clarify that the primary purpose of this PR is to close the Direct Messages view when the user navigates away. As per the comment from Tim [here](https://github.com/zulip/zulip/pull/28325#issuecomment-1897561567)

alya replied:
#27698 does not request any changes to when the DMs section collapses, which is consistent with @timabbott 's comment.

pratik-pc reviewed (commented):
Currently the highlight property of dm rows are being changed but the problem still remains that the conversation list doesn't update some of the time.

Rinwaoluwa commented:
@timabbott I've updated the PR with these changes as requested 

- RESOLVED CONFLICTS (rebased onto main)
- Removed white spaces and wrote a comment explaining is_private_messages_in_view  in handle_narrow_activated
- Updated commit message to mention the right function unhighlight_dm_list_item instead of deselect_dm_list_item

alya commented:
Closing this PR, as it sounds to me from the discussion above that this is based on a misunderstanding of the issue reported in https://github.com/zulip/zulip/issues/27698. Thanks for giving this a try, @Rinwaoluwa !

Rinwaoluwa commented:
That's a Bummer it was fun working on this I learnt a lot. Thank you.
